Allow sessions to be distinguished based on XRDP instance_name configuration by using new policy

lcniel opened the pull request:
See discussion: https://github.com/neutrinolabs/xrdp/discussions/3521

The motivation behind this PR is to allow a user (or an automation script) to maintain multiple distinct sessions based on the port configuration used when the session is started up. The reason for this is to allow xrdp to imitate the behaviour of multiple VNC servers started by users, while allowing a single instance of `xrdp-sesman` to track the sessions, and letting users (or middleware) choose which session to connect to based on a tag, instance_name, specified during startup, which can be equal to the port configuration, or some other string of choice that matches a particular `xrdp` configuration.

The specific use-case I have for this integrating `xrdp` with e.g., [OpenOnDemand](https://github.com/OSC/ondemand) to allow users to start multiple remote desktop sessions without having to worry about breakage if two sessions by coincidence start on the same compute node.

In my testing so far, using the  new `N` policy works just as well as the `D` policy - one has to ensure DBUS and XDG_RUNTIME_DIR don't get in the way of multiple xfce4 sessions, but the distinction of sessions by xrdp port works just fine.

This is just my suggestion for how to implement it, I expect that it will have to go through a few rounds of reviewing and editing.

edit: minor changes to reflect switch to instance_name field

matt335672 commented:
Thanks for this @lcniel - looks good.

I'm a bit snowed under with stuff IRL at the moment, but I hope to look at this early next week.

If you rebase to the latest devel when we've had a review, it will pull in #3684 to fix the FreeBSD CI issue.

lcniel commented:
> Thanks for this @lcniel - looks good.
> 
> I'm a bit snowed under with stuff IRL at the moment, but I hope to look at this early next week.
> 
> If you rebase to the latest devel when we've had a review, it will pull in #3684 to fix the FreeBSD CI issue.

Thanks, no rush at all. Hopefully I will be able to contribute a bit to the project moving forward since we're planning to use it quite a bit at my job.

matt335672 commented:
On a personal note, I got into xrdp coding because I was the UNIX administrator for a network where it was the best solution for what we were doing.  So I'm more that happy to help you with this.

lcniel commented:
I added the option to specify the port via xrdp-sesrun which I think makes sense for some use-cases - one may wish to start up such a port-bound session at a specific point in time, e.g. to make sure it ends up in the right cgroup.

matt335672 commented:
So sorry @lcniel - we're having to cope with a major family event here IRL.

I'll get to this if I get a chance, but it's unlikely to be until the new year now.

lcniel commented:
> So sorry @lcniel - we're having to cope with a major family event here IRL.
> 
> I'll get to this if I get a chance, but it's unlikely to be until the new year now.

No rush and no need to apologize! Please take care of yourself and your family.

lcniel commented:
Since I was making another PR, I took the opportunity to rebase this one against `devel`.

matt335672 commented on docs/man/sesman.ini.5.in:
```diff
@@ -235,6 +235,8 @@ Alternatively combine one-or-more of the following options
 .HP 4
 \fBI\fR - Sessions are separated by IP address
+.HP 4
+\fBI\fR - Sessions are separated by port configuration of XRDP server
```
'I' above should be a 'P'

lcniel replied:
Addressed.

lcniel replied:
OBE by change to instance_name implementation

matt335672 commented on libipm/ercp.h:
```diff
@@ -172,6 +172,7 @@ ercp_msg_in_reset(struct trans *trans);
  * @param guid Session GUID
  * @param start_ip_addr Starting IP address of client
  * @param start_time Session start time
+ * @param port Port listend to by xrdp client
```
'listend' -> 'listened'

lcniel replied:
Addressed.

lcniel replied:
OBE by change to instance_name implementation

matt335672 commented on sesman/scp_list.h:
```diff
@@ -84,6 +84,7 @@ struct scp_list_item
     uid_t uid; ///< User
     char *username; ///< Username from UID (at time of logon)
     char start_ip_addr[MAX_PEER_ADDRSTRLEN];
+    char xrdp_listen_port[1024]; ///< Port that xrdp listens on
```
I can see why there's a number here, but now with maintenance, we've introduced a dependency on a size match with the definition of port in `struct xrdp_startup_params`. Can we factor the 1024 out to a definition in `common/xrdp_constants.h`? Something like `MAX_XRDP_LISTEN_PORTLEN`.

lcniel replied:
Addressed.

lcniel replied:
Note that this field is now removed with the switch to instance_name and thus this suggestion is OBE.

matt335672 commented on sesman/session_list.h:
```diff
@@ -71,6 +71,10 @@ struct session_item
+    /* allow a user to run multiple sessions
+     * by running multiple instances of xrdp
+     * listening to different ports */
+    char xrdp_listening_port[1024];
```
Ditto for above comment

lcniel replied:
See above.

lcniel replied:
OBE per above.

matt335672 reviewed (commented):
Sorry about the delay. Here are some initial comments.

matt335672 commented:
Thanks @lcniel 

Owing to a slight issue we have at the moment, I'm unable to authorise your workflow above. Hopefully this will be sorted soon.

I can see why you've got this use-case and I'm happy to support it, but it's not going to be a common one I think. I'm wondering about hiding some of it for the more usual use-case of a single xrdp instance. Off the top of my head, there would be a couple of ways to do this.

1) In xrdp.ini, as well as specifying `port=`, specify an xrdp instance name to pass to sesman.ini (default blank). We could then use the instance name as the identifier rather than the raw port name. This is a bit of a design change, I admit, but might be a bit cleaner for the users. So, for example, you could have two xrdp instances running with instances of `DEVELOPMENT` and `SYSADMIN` (I'm making this up as I go along). These are the strings which would appear in the user interfaces, rather than the port names.
2) If the sesman policy doesn't contain a 'P', sesman can store the port names (in case of a restart), but simply return "", as the port name is immaterial.

There are other ways too. Please have a think about it, in particular the potential design change, and let me know your thoughts.

lcniel commented:
Thanks a lot for the feedback!

> Thanks @lcniel
> 
> Owing to a slight issue we have at the moment, I'm unable to authorise your workflow above. Hopefully this will be sorted soon.
> 
> I can see why you've got this use-case and I'm happy to support it, but it's not going to be a common one I think. I'm wondering about hiding some of it for the more usual use-case of a single xrdp instance. Off the top of my head, there would be a couple of ways to do this.

Yeah, I completely agree, it's an edge case even for me and because I want to avoid doing hacks with sending different initial resolutions and use something more intuitive.

>     1. In xrdp.ini, as well as specifying `port=`, specify an xrdp instance name to pass to sesman.ini (default blank). We could then use the instance name as the identifier rather than the raw port name. This is a bit of a design change, I admit, but might be a bit cleaner for the users. So, for example, you could have two xrdp instances running with instances of `DEVELOPMENT` and `SYSADMIN` (I'm making this up as I go along). These are the strings which would appear in the user interfaces, rather than the port names.

This should work just fine and is actually a feature with a wider and potentially more useful scope than my approach, since it covers more use-cases. So let's say a user has two jobs (job1 and job2) which both start up a user-scoped xrdp, each of which listens to a randomized port number. The first xrdp starts listening to port X with instance name job1, the other starts listening to port Y with instance name job2.  User connects to port X and therefore to the fist xrdp instance which has the name job1, and thus gets forwarded to the session associated with job1. Ditto for job2 and port Y. As you say it could also be useful for administrators, development and similar. As long as it still works with xrdp-sesrun specifying the tag it's fine with me. I think apart from renaming the only functionality change will be that in `xrdp_mm.c` the variable to be sent will be fetched from a different field read from the config file, instead of being the literal port.

>     2. If the sesman policy doesn't contain a 'P', sesman can store the port names (in case of a restart), but simply return "", as the port name is immaterial.
> 
> 
> There are other ways too. Please have a think about it, in particular the potential design change, and let me know your thoughts.

I don't have any problem with this but it's not totally clear to me what kind of design change this would imply.


I will address your PR comments tomorrow and if I have time also draft an implementation of your suggestion 1 above.

lcniel commented:
@matt335672 I have updated the PR to reflect your suggestion 1, using instance_name instead. It works just fine in my testing, so I don't think it makes sense to make a parallel PR for comparison, I don't see any reason not to use this approach.

matt335672 commented on sesman/scp_list.h:
```diff
@@ -84,6 +84,7 @@ struct scp_list_item
     uid_t uid; ///< User
     char *username; ///< Username from UID (at time of logon)
     char start_ip_addr[MAX_PEER_ADDRSTRLEN];
+    char xrdp_instance_name[MAX_XRDP_INSTANCE_NAMELEN]; ///< Port that xrdp listens on
```
Comment needs fixing.

lcniel replied:
Done.

matt335672 commented on xrdp/xrdp.c:
```diff
@@ -302,6 +302,12 @@ read_xrdp_ini_startup_params(struct xrdp_startup_params *startup_params)
+            if (g_strcasecmp(name, "instance_name") == 0)
+            {
+                g_strncpy(startup_params->instance_name, val,
+                          sizeof(startup_params->instance_name) - 1);
```
For new code, prefer `strlcpy()` which is supported in devel stream.

```c
                strlcpy(startup_params->instance_name,
                        sizeof(startup_params->instance_name), val);
```

`g_strncpy()` is not an efficient function!

lcniel replied:
Done.

matt335672 commented on common/xrdp_constants.h:
```diff
@@ -43,6 +43,12 @@
+ * Max length of buffer containing an instance name used to
+ * tag sessions for discrimination on reconnection.
+ **/
+#define MAX_XRDP_INSTANCE_NAMELEN 1024
```
Do we need this length? 80 should be enough(?)

lcniel replied:
Agreed. I originally set it to match the length of the port field but at this point I imagine I'd just use a hash of that or something instead of the literal exact string. Addressed and changed to 80.

matt335672 reviewed (commented):
Looking very nice!

Some more minor comments below.

Also, `instance_name` needs documenting in xrdp.ini and the xrdp.ini manpage.

lcniel commented:
> Looking very nice!
> 
> Some more minor comments below.
> 
> Also, `instance_name` needs documenting in xrdp.ini and the xrdp.ini manpage.

I have updated the documentation as well now.

matt335672 commented on docs/man/xrdp.ini.5.in:
```diff
@@ -28,6 +28,15 @@ The options to be specified in the \fB[Globals]\fR section are the following:
+.TP
+\fBinstance_name\fR=\fI<arbitrary-string>\fR
+Specify an arbitrary name for the xrdp instance that can be used to
+identify sessions that this xrpd instance started, or which were
```
xrpd -> xrdp

lcniel replied:
Fixed.

matt335672 reviewed (commented):
Typo.

matt335672 commented:
Looks good to me.

When I'm able, I'll be happy to merge this.

The code looks pretty clean to me, so now we've done the reviews, I'd say it's worth squashing to a single commit.

lcniel committed:
Introduce instance_name field into xrdp.ini and xrdp-sesrun, along with
the N policy in sesman.ini, allowing xrdp sessions to be tagged with an
instance name to enable persistent association with a specific
xrdp instance, to allow experiences where users reconnect to specific
sessions based on e.g. the xrdp listening port used.

lcniel commented:
> Looks good to me.
> 
> When I'm able, I'll be happy to merge this.
> 
> The code looks pretty clean to me, so now we've done the reviews, I'd say it's worth squashing to a single commit.

Thanks a bunch, I have squashed and pushed. No issues at all.
